### PR TITLE
Merge 3.1

### PR DESCRIPTION
**Pull request by wallyworld**
Merge 3.1

https://github.com/juju/juju/pull/16598 [from wallyworld/get-owned-secrets](https://github.com/juju/juju/commit/4d0cd7d25b107f85a89fbff222f75600a204d199)
https://github.com/juju/juju/pull/16596 [from manadart/2.9-fix-bad-clouds-panic](https://github.com/juju/juju/commit/7027d7920051b5c1b44a951d94c004fa1a53932c)
https://github.com/juju/juju/pull/16603 [from wallyworld/inhook-secret-refresh](https://github.com/juju/juju/commit/cb7ad3b64a683f2453d6868a1e570ec4c791c793)
https://github.com/juju/juju/pull/16583 [from jack-w-shaw/update_merge_bot](https://github.com/juju/juju/commit/d4b9e0021bf2a6dee6ad9bac68aeb8b54cc7fae2)
https://github.com/juju/juju/pull/16584 [from hmlanigan/fix-deploy-ci-lxd-series](https://github.com/juju/juju/commit/7c1d9951264713557340cb3fee25df7dcb579519)
https://github.com/juju/juju/pull/16618 [from manadart/2.9-destroy-storage-with-test…](https://github.com/juju/juju/commit/3b81a0023b9f1efc48e3ede8479a6a0b75b2c29e)
https://github.com/juju/juju/pull/16617 [from wallyworld/remove-placeholder-port](https://github.com/juju/juju/commit/d57438acebf9a47d99d665db57fd2be098127086)
https://github.com/juju/juju/pull/16616 [from wallyworld/fix-topologykey-constraint](https://github.com/juju/juju/commit/060e1756f64344cfe408a0989a7f838d40a28fa9)
https://github.com/juju/juju/pull/16465 [from hpidcock/fix-docker-auth-challenges](https://github.com/juju/juju/commit/f80096d78ffd95291aad561fddc2fd436638deae)
https://github.com/juju/juju/pull/16610 [from wallyworld/cli-help-fixes](https://github.com/juju/juju/commit/dc94b094ecb853c8f4c50095da3ffdd3ee7b45d3)
https://github.com/juju/juju/pull/16609 [from SimonRichardson/fix-panic](https://github.com/juju/juju/commit/5fcd0f69f255c2b1b0f1b1bb255746e520fc4b1c)
https://github.com/juju/juju/pull/16576 [from SimonRichardson/pre-check-facades-migr…](https://github.com/juju/juju/commit/ad85101acace0417da97a5405d3a78aa3be44f67)

Conflicts
```
# Conflicts:
#       api/package_test.go
#       apiserver/allfacades.go
#       cmd/juju/application/integrate.go
#       cmd/juju/model/defaults.go
#       core/charm/repository/charmhub.go
#       core/charm/repository/charmhub_test.go
#       go.mod
#       go.sum
#       state/secrets.go
#       state/upgrades.go
#       state/upgrades_test.go
#       tests/suites/deploy/deploy_bundles.sh
#       tests/suites/secrets_iaas/juju.sh
#       tests/suites/secrets_iaas/vault.sh
#       upgrades/backend.go
#       upgrades/steps_317.go
#       upgrades/steps_317_test.go
```

### CODE REVIEW

File: core/charm/repository/charmhub.go [hunk 1]
```diff
@@ -109,7 +109,14 @@ func (c *CharmHubRepository) ResolveForDeploy(arg corecharm.CharmID) (corecharm.
...
 //     charm downloads later.
-func (c *CharmHubRepository) resolveWithPreferredChannel(charmName string, requestedOrigin corecharm.Origin) (*charm.URL, corecharm.Origin, []corecharm.Platform, transport.RefreshResponse, error) {
+func (c *CharmHubRepository) resolveWithPreferredChannel(charmName string, argOrigin corecharm.Origin) (*charm.URL, corecharm.Origin, []corecharm.Platform, transport.RefreshResponse, error) {
+	c.logger.Tracef("Resolving CharmHub charm %q with origin %v", charmName, argOrigin)
+
+	requestedOrigin, err := c.validateOrigin(argOrigin)
+	if err != nil {
+		return nil, corecharm.Origin{}, nil, transport.RefreshResponse{}, errors.Trace(err)
+	}
+
```
Comment (hmlanigan) [inline core/charm/repository/charmhub.go:119]: Adding the logger trace is good. This bit of code isn't needed in 3.3, it got moved to ResolveWithPreferredChannel().

File: tests/suites/deploy/deploy_charms.sh [hunk 2]
```diff
@@ -14,6 +14,23 @@ run_deploy_charm() {
 	destroy_model "test-deploy-charm"
 }
 
+run_deploy_charm_unsupported_series() {
```
Comment (hmlanigan) [inline tests/suites/deploy/deploy_charms.sh:17]: nice, once this lands I can re-generate the tests for Jenkins and get more green in proving grounds.

File: tests/suites/secrets_iaas/vault.sh [hunk 3]
```diff
@@ -5,8 +5,7 @@ run_secrets_vault() {
 
 	model_name='model-secrets-vault'
 	juju add-secret-backend myvault vault endpoint="$VAULT_ADDR" token="$VAULT_TOKEN"
-	add_model "$model_name"
-	juju --show-log model-config secret-backend=myvault -m "$model_name"
+	add_model "$model_name" --config secret-backend=myvault
```
Comment (ycliuhw) [inline tests/suites/secrets_iaas/vault.sh:9]: This was a fix I added we should keep.
The `add-model` does not support extra args for model config.
Ah, we actually have to backward port these changes to 3.1.

**Review by ycliuhw — APPROVED** [review]

**Pull request merged**